specs/plasma.md

```diff
@@ -0,0 +1,114 @@
+
+- A Commitment type (i.e. keccak256)
+- Commitment bytes
+- The L1 block number in which it was included.
+
+Users with access to the input data then have another window of L1 blocks (AKA `resolve_window`)
+during which they can submit it as calldata to the chain by calling the `resolve` method of the contract.
+If the data is not included onchain by the time the resolve window is elapsed, derivation of the L2 canonical chain
+will omit the input data, hence any transaction included in the input data will be dropped
+causing a reorg of any L2 chain that previously had included that transaction.
+
+In order to challenge a commitment, users must deposit a bond amount where `bond >= resolve_tx_gas_cost`.
+If the gas cost of resolving the challenge was lower than the bond, the difference is reimbursed to the challenger
+and the rest of the bond is burnt. If the challenge is not resolved in time and expired,
+the bond is returned and can be withdrawn by the challenger or used to challenge another commitment.
+
+The state of all challenges can be read from the contract state or by syncing contract events.
+`challenge_window` and `resolve_window` are constant values that currently cannot be changed
+unless the contract is upgraded. A dynamic window mechanism may be explored in the future.
+
+## Derivation
+
+Input data is retrieved during derivation of L2 blocks. The changes to the derivation pipeline
+when using the alt DA source are limited to swapping out the L1 based `DataAvailabilitySource` step
+in the pipeline with a module that enables pulling the data from the offchain DA source.
+
+Similarly to L1 based DA, for each L1 block we open a calldata source to retrieve the input commitments
+from the transactions and use each commitment with its l1 origin block number to resolve
+the input data from the storage service.
+
+In addition, we filter events from the DA Challenge contract included in the block
+and sync a local state of challenged input commitments. As the derivation pipeline steps through
+`challenge_window + resolve_window` amount of L1 blocks, any challenge marked as active
+becomes expired causing a reset of the derivation pipeline.
+
+Derivation can either be driven by new L1 blocks or restarted (reset) from the L1 origin of the last
+L2 safe head known by the execution engine.
+
+When stepping through new L1 origin blocks, input data is loaded from the DA storage service.
+If the service responds with a 404 (not found) error, derivation stalls and the DA Manager
+starts a detached traversal of new L1 blocks until:
+
+- An `Active` challenge event is found, detached L1 traversal continues:
+  - A `Resolved` challenge event is found: derivation continues with the input data submitted onchain.
+  - A high enough L1 block is reached such as `latest_l1 - pipeline_l1_origin = resolve_window_size`:
+    input data is skipped from the derivation output.
+- Data is never challenged, derivation returns critical error.
```

Comment:
This opens up the potential to break nodes by providing the data to a DA storage service initially, then after the challenge period elapses, the data is removed/lost from the DA.  That's not unreasonable and aligns with how 4844 works, but I wonder if we need to be explicit about this form of weak subjectivity in the specs somewhere. Essentially your node can't reliably sync from L1 if it is more than the challenge window behind the current chain head.

I'm not sure exactly how that should be called out, so mostly just flagging this as a question - hopefully the folks who've been working on 4844 can chime in.

Reply:
This sort of attack only occurs if the user doesn't have an honest peer to connect to on the p2p network. This user should be able to both full sync and snap sync. The EL will index the blocks, which are approximately the inputs to the L2. The assumption is that users will download the state within the challenge window if they care about the chain. It is also permissionless for an alternative DA service to pull in and index the data to make another DA store

Reply:
The EL will index blocks, but it's not enough to go back to batches at this point (though I'm interested in exploring that space). I agree I'd like to see a section of weak subjectivity / what happens when the challenge window passes.